### ISSUE TITLE
IKE/CHILD SA not deleted after --load-conn config changes.

### ISSUE THREAD

**Issue by NLako-6WIND** [strongswan/strongswan#2324]
**System (please complete the following information):**
 - OS:  Ubuntu 22.04
 - Kernel version: 5.15.0-113-generic
 - strongSwan version(s): 5.9.13
 - Tested/confirmed with the latest version: no (but code seemingly didn't change)

**Describe the bug**

**1** - When changing IKE_SA parameters by loading a new peer_config
    configuration through 'swanctl --load-conn', if the related IKE_SA had
    more than 2 CHILD and all of them had 'start' as their 'start_action',
    those children would be deleted, and re-initiated in a new IKE_SA, but
    the 'old' IKE_SA would not be deleted.
    **1.1** - This seems un-intended as if the IKE_SA has between 0 and 2 (included)
    CHILD (all with 'start' for 'start_action'), the 'old' IKE_SA would be deleted
    **1.2** - Finaly the IKE_SA would be deleted at the reauth time, but it could be rekeyed
    several time before this. It poses some constrains when dealing with a big number
    of IKE_SAs and HA at the same time.

**2** - When changing child_cfg(s) by using the same 'swanctl --load-conn' command:
    If the IKE SA had two child with 'start_action=start', and both of the related child_cfg
    were changed. The IKE was deleted, but it's children not re-initiated.
    **2.1** - In case if the IKE had only one children, and the latest child_cfg changed, the IKE
    would be deleted, and the child re-initiated within a new IKE_SA ; In case if the IKE had
    >= 3 children, (with all of them changed), children would be deleted and re-initiated within
   the same IKE.

**To Reproduce**
1. Use two hosts connected to each other (in my case, two VM's)
2. On host1:
```shell
for i in 11 12 13 21 22 23 18 ; do ip link add loop-$i type dummy ; ip addr add 192.168.10.$i dev loop-$i ; ip link set dev loop-$i up ; done
ip addr add 192.168.10.1/32 dev $connected_interface
ip route add 192.168.10.0/26 dev $connected_interface
```
  Then upload the related host1_swanctl.conf (in attachements) and copy it inside your swanctl.conf

3. On host2:
```shell
for i in 41 42 43 31 32 33 48 ; do ip link add loop-$i type dummy ; ip addr add 192.168.10.$i dev loop-$i ; ip link set dev loop-$i up ;  done
ip addr add 192.168.10.2/32 dev $connected_interface
ip route add 192.168.10.0/26 dev $connected_interface
```
  Then upload the related host2_swanctl.conf (in attachements) and copy it inside your swanctl.conf 
4. copy content of files in attachements in their respective swanctl.conf
5. (Re)Start strongswan.service on both hosts
6.  On host1 - manually initiate remaining childs:
```shell
for i in nonee-child-301 trapp-child-301 ; do swanctl --initiate --child $i ; done
``` 
7. Then replace ike rekey/reauth times for first three vpns in the config, and the rekey_time for all others vpns 'childs':
```shell
sed -i '150,352s/^[ ]\{8\}rekey_time =.*$/        rekey_time = 1h/' $PATH_TO_SWANCTL_CONF
sed -i '6,149s/^[ ]\{4\}rekey_time =.*$/    rekey_time = 10h/' $PATH_TO_SWANCTL_CONF
sed -i '6,149s/^[ ]\{4\}reauth_time =.*$/    reauth_time = 100h/' $PATH_TO_SWANCTL_CONF
``` 
8. Load connections:
```shell
swanctl --load-conn
``` 

** Current behavior **
(1) duplicate of a IKE_SA
> root@dut-vm:~# swanctl --list-sas --ike vpn_four_start | grep -e \\#
> vpn_four_start: #10, ESTABLISHED, IKEv2, 8a6c378aaae274ff_i* e6a3f46f3434d39b_r
>   start-child-021: #21, reqid 1, INSTALLED, TUNNEL, ESP:AES_GCM_16-256
>   start-child-022: #25, reqid 8, INSTALLED, TUNNEL, ESP:AES_GCM_16-256/CURVE_25519
>   start-child-023: #27, reqid 7, INSTALLED, TUNNEL, ESP:AES_GCM_16-256/CURVE_25519
>   start-child-024: #29, reqid 9, INSTALLED, TUNNEL, ESP:AES_GCM_16-256/CURVE_25519
> vpn_four_start: #3, ESTABLISHED, IKEv2, 93ee01ff7e73f79e_i* 33086f3d90544195_r

(2) related children didn't execute their START start_action proprely, they're missing
> root@dut-vm:~# swanctl --list-sas --ike vpn_two_start_chld_test
> root@dut-vm:~# 

**Expected behavior**

(1)  vpn_four_start is present only 1 time:
> 
> root@dut-vm:~# swanctl --list-sas --ike vpn_four_start | grep -e \\#
> vpn_four_start: #10, ESTABLISHED, IKEv2, 8a6c378aaae274ff_i* e6a3f46f3434d39b_r
>   start-child-021: #21, reqid 1, INSTALLED, TUNNEL, ESP:AES_GCM_16-256
>   start-child-022: #25, reqid 8, INSTALLED, TUNNEL, ESP:AES_GCM_16-256/CURVE_25519
>   start-child-023: #27, reqid 7, INSTALLED, TUNNEL, ESP:AES_GCM_16-256/CURVE_25519
>   start-child-024: #29, reqid 9, INSTALLED, TUNNEL, ESP:AES_GCM_16-256/CURVE_25519

(2)  child_sa should be restarted, initiating the IKE
> root@dut-vm:~# swanctl --list-sas --ike vpn_two_start_chld_test | grep -e \\#
> vpn_two_start_chld_test: #5, ESTABLISHED, IKEv2, 0835456b9a4940b6_i* b2f54bfeeba1f224_r
>  start-child-111: #8, reqid 8, INSTALLED, TUNNEL, ESP:AES_GCM_16-256
>  start-child-112: #13, reqid 12, INSTALLED, TUNNEL, ESP:AES_GCM_16-256/CURVE_25519

**Logs/Backtraces**
If applicable, add logs or backtraces to help explain your problem.

**Additional context**

Initialy, only the (1) issues was seen, after further testing the (2) appeared, and also I've found a third one
(If the IKE SA had children with diferent 'start_action' configured, changing child_cfg resulted in 'duplicates'
of some 'start_action=start' children).
I've made a draft patch that resolved the issues for me. Commit log contains some more details on the
potential 'root-cause'

[host1_swanctl.conf.txt](https://github.com/user-attachments/files/16088099/host1_swanctl.conf.txt)
[host2_swanctl.conf.txt](https://github.com/user-attachments/files/16088100/host2_swanctl.conf.txt)
[0001-vici-clean-IKE_SAs-after-loading-new-connections-con.patch.txt](https://github.com/user-attachments/files/16088102/0001-vici-clean-IKE_SAs-after-loading-new-connections-con.patch.txt)

**Comment by tobiasbrunner** [comment]
Thanks for the detailed report.

As you noticed, one issue is that with multiple CHILD_SAs, there is a race condition as they are not immediately removed but go through a proper DELETE cycle. So because the current code only looks at a single child config at a time, IKE_SAs might appear to have "other", i.e. unrelated, CHILD_SAs and it won't get deleted.

Another issue is that when re-initiating a CHILD_SA, there could be another race condition because the task to delete the IKE_SA is queued like the CHILD_SA deletions. While IKE_SA deletions are preferred over CHILD_SA deletions when activating new tasks, if one of the latter (or any other task) is already active, the IKE_SA will look like it is still usable when initiating new CHILD_SAs, while it is actually not. The queued tasks to create CHILD_SAs will then be lost once the task to delete the IKE_SA is activated. We can add a check to `ike_sa_manager_t::checkout_by_config()` that looks at the queued tasks to avoid using such IKE_SAs. But yeah, we should probably avoid deleting IKE_SAs in the first place if only the child configs have changed and there are CHILD_SAs to be initiated.

While we could add a separate step to clean up the IKE_SAs after going through all the child configs, I'd like to avoid that. It requires enumerating the IKE_SAs yet another time, in addition to the enumerations the current code already requires because each child config is handled separately, which is not ideal in the first place. Besides that, your patch has other issues, for instance, it doesn't handle the `unload-conn` case and locking and the `handling_actions` flag are also not treated correctly.

I feel we could make the cleanup more efficient by enumerating the IKE_SAs only once for all the removed child configs. Then we'd also know if we are able to remove the IKE_SA or not. I've pushed some prototypical changes to the `2324-reload-actions` branch, which should address the issues. Please let me know what you think.

**Comment by tobiasbrunner** [comment]
Thanks for testing. And yes you're absolutely right, it should be in the `if` block, not the `else` block. I've updated the branch.

**Comment by NLako-6WIND** [comment]
Thanks for everything,

Can I ask, when could we except to see the ```2324-reload-actions``` branch to be merged in master?

**Comment by tobiasbrunner** [comment]
Probably in the coming weeks.

**Comment by NLako-6WIND** [comment]
Thanks very much :)
